src/optimizers/bff_optimizer.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# BFF_Optimizer: a pure Bfloat16 AdamW optimizer with optional Kahan summation
+# Features direct control over momentum, variance and auxiliary compensation
+# buffer dtypes.
+# Kahan summation is used to offset Bfloat16 precision reduction for
+# the weight updates, allowing full training in BFloat16.
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class BFF_Optimizer(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        betas=(0.9, 0.999),
+        eps=1e-8,
+        weight_decay=0.0,
+        use_kahan_summation=True,
+        # use_matching_params_dtype=False,
+        momentum_dtype=torch.bfloat16,
+        variance_dtype=torch.bfloat16,
+        compensation_buffer_dtype=torch.bfloat16,
+    ):
+        """
+        Args:
+                params (iterable): iterable of parameters to optimize or dicts defining
+                    parameter groups
+                lr (float, optional): learning rate (default: 1e-3)
+                betas (Tuple[float, float], optional): coefficients used for computing
+                    running averages of gradient and its square (default: (0.9, 0.999))
+                eps (float, optional): term added to the denominator to improve
+                    numerical stability (default: 1e-8)
+                weight_decay (float, optional): weight decay coefficient (default: 1e-2)
+
+                # BFF specific
+                use_kahan_summation = creates auxiliary buffer to ensure high precision
+                model param updates
+                momentum_dtype = dtype for momentum
+                variance_dtype = dtype for uncentered variance
+                compensation_buffer_dtype = dtype for Kahan summation buffer
```

Comment:
maybe include a short example of how to use such an optimizer (i.e. does the user need to enable mixed precision? amp?) as part of the docstring.

Reply:
added